cpp/ql/lib/semmle/code/cpp/exprs/Expr.qll

```diff
@@ -1015,8 +1015,19 @@ class DeleteOrDeleteArrayExpr extends Expr, TDeleteOrDeleteArrayExpr {
   Expr getExpr() {
     // If there is a destructor call, the object being deleted is the qualifier
     // otherwise it is the third child.
-    result = this.getChild(3) or result = this.getDestructorCall().getQualifier()
+    exists(Expr exprWithReuse | exprWithReuse = this.getExprWithReuse() |
+      if not exprWithReuse instanceof ReuseExpr
+      then result = exprWithReuse
+      else result = this.getDestructorCall().getQualifier()
+    )
   }
+
+  /**
+   * Gets the object or array being deleted, and gets a re-use expression when
+   * there is a destructor call and the object is also the qualifier of the
+   * call.
+   */
```

Comment:
Sure. Suggestions?

Reply:
Maybe something like:
```ql
/**
 * Gets the object or array being deleted, and gets a re-use expression when
 * there is a destructor call and the object is also the qualifier of the
 * call.
 * 
 * For example, given:
 * ```
 * struct HasDestructor { ~HasDestructor(); };
 * struct PlainOldData { int x, char y; };
 * void f(HasDestructor* hasDestructor, PlainOldData* pod) {
 *   delete hasDestructor;
 *   delete pod;
 * }
 * ```
 * the query:
 * ```
 * from DeleteExpr del
 * where del.getExprWithReuse() instanceof ReuseExpr
 * select del
 * ```
 * selects `delete hasDestructor;` since the deleted expression has a destructor,
 * and the deleted expression is a "reuse" of the expression that was the qualifier
 * of the destructor call.
 */
```
?

Reply:
Added something slightly more compact.